README
dvb-vulture
===========

This Archive contains the DVB Vulture, which
is a DVB streaming software with features. 

What it can do:
* RTP streaming. Multiple services possible.
* Recording. both timed and unconstrained. Multiple services possible.
* Few library/software dependencies.
* The libraries it does depend on are rather mature(=old) and stable.
* Automated service listing and selection. Never again dvbscan and crappy channel list text files.
* Background service/transponder list updates.
* System clock update.
* Stream announcement(using SAP/SDP)
* EPG schedule collection(+Viewer)
* Videotext collection(+Viewer)
* DVB-S and DVB-T support(DVB-S2 is untested and may not work)
* Complex switch/antenna Setups
* Follows pid changes by tracking PMT.
* Reference counting software demultiplexer, no copying of packets.
* Basic multi-user arbitration.

What it can _not_ do:
- Pay TV(Not even with CI and card)
- Play streams or files. But vlc or xine work well.
- Hbb stuff(there's a downloader that I nicked from Simon Kilvingtons redbutton software, but no viewer)

It is licensed under the GPLv3.

Build Configuration
-------------------

There is no configure script. 
If you want to change something, you will have to edit
the makefiles yourself. 

You can override some make variables: 

`prefix` defaults to `/usr/local`
`exec_prefix` defaults to  `$(prefix)`
`datarootdir` defaults to `$(prefix)/share`
`bindir` defaults to `$(exec_prefix)/bin`
`infodir` defaults to `$(datarootdir)/info`
`libdir` defaults to `$(exec_prefix)/lib`

Installation
------------

Just `make install` in the top level source directory, 
to make and install all necessary files (executable and .info manuals). 
It is also possible to create a .deb binary package, 
using `dpkg-buildpackage -uc -us -b` here. Parallel builds should work, 
too: `dpkg-buildpackage -uc -us -b -j20`.

Docs
----

The man page is just a stub. Documentation is maintained in texinfo format.
Texinfo pages can be viewed using `info dvbv-stream`
or `info dvbv-nr`. 

Dmalloc
-------

When compiling with `-DDMALLOC_DISABLE` make sure to 
use it for _all_ of the programs as well as the _library_. 
Otherwise there _will_ be segfaults. Background: pointers to dynamically 
allocated blocks of memory are passed between program and library. 
Some are freed/allocated on either side. 

Library dependencies(no need for the exact same versions, though)
----------------------------------------------------------------

dvbv-stream:

	linux-gate.so.1 =>  (0xb783f000)
	libdmallocth.so.5 => /usr/lib/libdmallocth.so.5 (0xb77a6000)
	libpthread.so.0 => /lib/i386-linux-gnu/libpthread.so.0 (0xb778d000)
	libsacommon.so => /usr/lib/libsacommon.so (0xb7770000)
	libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xb762b000)
	/lib/ld-linux.so.2 (0xb7840000)

dvbv-nr:

	linux-gate.so.1 =>  (0xb784f000)
	libsacommon.so => /usr/lib/libsacommon.so (0xb7819000)
	libdmallocth.so.5 => /usr/lib/libdmallocth.so.5 (0xb7799000)
	libncurses.so.5 => /lib/libncurses.so.5 (0xb7760000)
	libmenu.so.5 => /usr/lib/libmenu.so.5 (0xb7759000)
	libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xb7614000)
	libdl.so.2 => /lib/i386-linux-gnu/libdl.so.2 (0xb760f000)

dvbv-xr:
	linux-gate.so.1 =>  (0xb7784000)
	libXext.so.6 => /usr/lib/i386-linux-gnu/libXext.so.6 (0xb7751000)
	libX11.so.6 => /usr/lib/i386-linux-gnu/libX11.so.6 (0xb7619000)
	libXt.so.6 => /usr/lib/i386-linux-gnu/libXt.so.6 (0xb75bc000)
	libXaw.so.7 => /usr/lib/libXaw.so.7 (0xb755f000)
	libsacommon.so => /usr/lib/libsacommon.so (0xb753f000)
	libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xb73f5000)
	libxcb.so.1 => /usr/lib/i386-linux-gnu/libxcb.so.1 (0xb73d2000)
	libdl.so.2 => /lib/i386-linux-gnu/libdl.so.2 (0xb73ce000)
	libSM.so.6 => /usr/lib/i386-linux-gnu/libSM.so.6 (0xb73c6000)
	libICE.so.6 => /usr/lib/i386-linux-gnu/libICE.so.6 (0xb73ad000)
	libXmu.so.6 => /usr/lib/i386-linux-gnu/libXmu.so.6 (0xb7393000)
	libXpm.so.4 => /usr/lib/i386-linux-gnu/libXpm.so.4 (0xb7382000)
	/lib/ld-linux.so.2 (0xb7785000)
	libXau.so.6 => /usr/lib/libXau.so.6 (0xb737f000)
	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0xb737a000)
	libuuid.so.1 => /lib/i386-linux-gnu/libuuid.so.1 (0xb7374000)


libsacommon.so:

	/lib/ld-linux.so.2 (0xb7850000)
	linux-gate.so.1 =>  (0xb77ca000)
	libdmallocth.so.5 => /usr/lib/libdmallocth.so.5 (0xb7714000)
	libc.so.6 => /lib/i386-linux-gnu/libc.so.6 (0xb75cf000)
	/lib/ld-linux.so.2 (0xb77cb000)

..and recently zlib(for the unfinished carousel stuff)

The source-code may be found at http://members.aon.at/~aklamme4/files.html

Note: the tab width should be set to 2 spaces when reading the source code. 
I tried to avoid spaces in indentation, but some may have crept in...


debian/changelog
dvbvulture (1.0.34) unstable; urgency=medium

  * Fix borked memsets
  * Minor Docs update

 -- Johann Klammer <klammerj@a1.net>  Mon, 28 Apr 2014 19:28:56 +0200

dvbvulture (1.0.33) unstable; urgency=medium

  * Fixed a race condition/segfault
  * minor menu and vt fixes

 -- Johann Klammer <klammerj@a1.net>  Sun, 20 Apr 2014 21:20:56 +0200

dvbvulture (1.0.32) unstable; urgency=medium

  * Fixes a segfault..

 -- Johann Klammer <klammerj@a1.net>  Tue, 08 Apr 2014 08:57:25 +0200

dvbvulture (1.0.31) unstable; urgency=medium

  * Fix recommends(so it actually builds)...

 -- Johann Klammer <klammerj@a1.net>  Mon, 07 Apr 2014 08:33:49 +0200

dvbvulture (1.0.30) unstable; urgency=medium

  * Disabled Videotext component, coz it crashes the server.
  * Proper charset conversion. (fresh and stinky, possibly broken)
  * Config file autogeneration.
  * Service tracking demuxer.
  * Signal monitor/status feedback hacks.
  * Loads of bugfixes.

 -- Johann Klammer <klammerj@a1.net>  Mon, 07 Apr 2014 02:05:30 +0200

dvbvulture (1.0.29) unstable; urgency=low

  * Added x_remote: Like ncurses_remote, but a Xaw application.
  * Added tsverify: Verify captured ts streams and output a lot of infos.
  * IPV6 support: Can do IPV4 or IPV6, but not both.
  * Bugfixes

 -- Johann Klammer <klammerj@a1.net>  Tue, 06 Nov 2012 20:40:55 +0100

dvbvulture (1.0.28) unstable; urgency=low

  * Fixed a Bug in Rec schedule.

 -- root <root@tivi>  Mon, 10 Sep 2012 07:48:49 +0200

dvbvulture (1.0.27) unstable; urgency=low

  * Parallel builds work now.
  * Better Inter Process Comunication.
  * Less compiler warnings.

 -- root <root@tivi>  Thu, 30 Aug 2012 22:13:22 +0200

dvbvulture (1.0.26) unstable; urgency=low

  * Recording schedule bugfix

 -- root <your.maintainer@address>  Sun, 17 Jun 2012 20:20:21 +0200

dvbvulture (1.0.25) unstable; urgency=low

  * Fixed a memleak
  * Added missing build-dep for libdmalloc-dev for dmalloc.h header and install-info.
  * Added dependency for info-browser

 -- root <your.maintainer@address>  Tue, 12 Jun 2012 03:40:46 +0200

dvbvulture (1.0.24) unstable; urgency=low

  * Minor bug fixes, cleanup.

 -- root <your.maintainer@address>  Mon, 04 Jun 2012 19:56:32 +0200

dvbvulture (1.0.23) unstable; urgency=low

  * Changed default install paths from /usr to usr/local.
    Various changes to makefile.
    Name of startup script changed to dvbvulture and moved to debian directory.
    Default state dir changed from /var/lib/stb to /var/lib/dvbvulture
    Default configuration dir changed from /etc/stb to /etc/dvbvulture
    Default local configuration dir changed from ~/.stb to ~/.dvbvulture
    Basic manpages for dvbvulture(pointing to real docs).

 -- root <your.maintainer@address>  Mon, 28 May 2012 20:04:08 +0200

dvbvulture (1.0.22) unstable; urgency=low

  * Initial debianization

 -- root <your.maintainer@address>  Thu, 24 May 2012 16:40:44 +0200


stream/eis.c
#include <stdlib.h>
#include <string.h>
#include <pthread.h>
#include "eis.h"
#include "size_stack.h"
#include "iterator.h"
#include "utl.h"
#include "debug.h"
#include "bits.h"
#include "item_list.h"
#include "timestamp.h"
#include "dmalloc.h"
int DF_EIS;
#define FILE_DBG DF_EIS
/*
a) To find Y, M, D from MJD
   Y' = int [ (MJD - 15 078,2) / 365,25 ]
   M' = int { [ MJD - 14 956,1 - int (Y' × 365,25) ] / 30,6001 }
   D = MJD - 14 956 - int (Y' × 365,25) - int (M' × 30,6001)
   If M' = 14 or M' = 15, then K = 1; else K = 0
   Y = Y' + K
   M = M' - 1 - K × 12
b) To find MJD from Y, M, D
   If M = 1 or M = 2, then L = 1; else L = 0
   MJD = 14 956 + D + int [ (Y - L) × 365,25] + int [ (M + 1 + L × 12) × 30,6001 ]
c) To find WD from MJD
   WD = [ (MJD + 2) mod 7 ] + 1
d) To find MJD from WY, WN, WD
   MJD = 15 012 + WD + 7 × { WN + int [ (WY × 1 461 / 28) + 0,41] }
e) To find WY, WN from MJD
   W = int [ (MJD / 7) - 2 144,64 ]
   WY = int [ (W × 28 / 1 461) - 0,0079]
   WN = W - int [ (WY × 1 461 / 28) + 0,41]
EXAMPLE:       MJD    =     45 218           W        = 4 315
               Y      =     (19)82           WY       = (19)82
               M      =     9 (September)    N        = 36
               D      =     6                WD       = 1 (Monday)
NOTE:   These formulas are applicable between the inclusive dates 1900 March 1 to 2100 February 28.
*/


int
get_eis_tsid (uint8_t * sec)
{
  return bitsGet (sec, 8, 0, 16);
}

static int
get_event (SizeStack * s, EVT * evt)
{
  uint8_t *ptr;
  uint16_t desc_len;
  uint16_t mjd;
  uint8_t h1, m1, s1, h2, m2, s2;

  if (sizeStackAdvance (s, 0, 12))
    return 1;
  ptr = sizeStackPtr (s);
  evt->evt_id = bitsGet (ptr, 0, 0, 16);

  mjd = bitsGet (ptr, 2, 0, 16);
  h1 = bitsDecodeBcd8 (bitsGet (ptr, 2, 16, 8));
  m1 = bitsDecodeBcd8 (bitsGet (ptr, 2, 24, 8));
  s1 = bitsDecodeBcd8 (bitsGet (ptr, 2, 32, 8));
  evt->start = get_time_mjd (mjd) + get_time_hms (h1, m1, s1);

  h2 = bitsDecodeBcd8 (bitsGet (ptr, 7, 0, 8));
  m2 = bitsDecodeBcd8 (bitsGet (ptr, 7, 8, 8));
  s2 = bitsDecodeBcd8 (bitsGet (ptr, 7, 16, 8));
  evt->duration = get_time_hms (h2, m2, s2);
  evt->running_status = bitsGet (ptr, 10, 0, 3);
  evt->ca_mode = bitsGet (ptr, 10, 3, 1);

  debugPri (2,
            "event id:%hu, start:%hhu:%hhu:%hhu, duration:%hhu:%hhu:%hhu, status:%hhu, ca:%hhu\n",
            evt->evt_id, h1, m1, s1, h2, m2, s2, evt->running_status,
            evt->ca_mode);
  desc_len = bitsGet (ptr, 10, 4, 12);
  if (sizeStackAdvance (s, 12, 0))
    return 1;
  if (sizeStackPush (s, desc_len))
    return 1;
  while (desc_len > 0)
  {
    uint8_t d_tag, d_len;
    if (sizeStackAdvance (s, 0, 2))
      return 1;
    ptr = sizeStackPtr (s);
    d_tag = ptr[0];
    d_len = ptr[1];
    if (sizeStackAdvance (s, 2, 0))
      return 1;
    if (sizeStackPush (s, d_len))
      return 1;
    ptr = sizeStackPtr (s);
    switch (d_tag)
    {
    case 0x4d:                 //short event desc
      if (!evt->name)
      {
        if (sizeStackAdvance (s, 0, 4))
          return 1;
        evt->name = utlCalloc (1, ptr[3] + 1);
        if (sizeStackAdvance (s, 0, 4 + ptr[3]))
          return 1;
        memmove (evt->name, ptr + 4, ptr[3]);
        evt->name[ptr[3]] = '\0';

        if (sizeStackAdvance (s, 0, 5 + ptr[3]))
          return 1;

        evt->description = utlCalloc (1, ptr[4 + ptr[3]] + 1);
        if (sizeStackAdvance (s, 0, 5 + ptr[3] + ptr[4 + ptr[3]]))
          return 1;
        memmove (evt->description, ptr + 5 + ptr[3], ptr[4 + ptr[3]]);
        evt->description[ptr[4 + ptr[3]]] = '\0';
      }
      else
        debugPri (2, "Ignoring short event desc\n");
      break;
//                      case 0x4e://extended event desc
      //              break;ignore for now
    case 0x54:                 //content desc
      if ((!evt->content1) && (!evt->content2))
      {
        evt->content1 = bitsGet (ptr, 0, 0, 4);
        evt->content2 = bitsGet (ptr, 0, 4, 4);
      }
      break;
    case 0x55:                 //parental rating desc
      if (!evt->rating)
        evt->rating = ptr[3];
      break;
    default:
      debugPri (2, "Event ignoring descriptor: %u\n", d_tag);
      break;
    }
    if (sizeStackPop (s))
      return 1;
    desc_len -= 2 + d_len;
    if (sizeStackAdvance (s, d_len, 0))
      return 1;
  }
  if (sizeStackPop (s))
    return 1;
  return 0;
}

int
parse_eis (uint8_t * eis, EIS * p)
{
  uint16_t section_length;
  uint16_t svid;
  uint16_t onid;
  uint16_t tsid;
  uint32_t crc;
  uint8_t version_number;
  uint8_t current_next;
  uint8_t section_number;
  uint8_t last_section_number;
  uint8_t segment_last_section_number;
  uint8_t last_table_id;
  SizeStack s;
  uint8_t *ptr;
  int i, j, k;


  if (sizeStackInit (&s, eis) || sizeStackPush (&s, MAX_SEC_SIZE))      //maximum section size
    return 1;

  if (sizeStackAdvance (&s, 1, 2))
    return 1;
  ptr = sizeStackPtr (&s);
  section_length = bitsGet (ptr, 0, 4, 12);

  if (sizeStackAdvance (&s, 2, 1))
    return 1;

  if (sizeStackPush (&s, section_length))
    return 1;

  if (sizeStackAdvance (&s, 0, 11))
    return 1;

  ptr = sizeStackPtr (&s);

  svid = bitsGet (ptr, 0, 0, 16);
  version_number = bitsGet (ptr, 2, 2, 5);
  current_next = bitsGet (ptr, 2, 7, 1);
  section_number = ptr[3];
  last_section_number = ptr[4];
  tsid = bitsGet (ptr, 5, 0, 16);
  onid = bitsGet (ptr, 7, 0, 16);
  segment_last_section_number = ptr[9];
  last_table_id = ptr[10];
  debugPri (1,
            "EIS: svid: %u, tsid: %u, onid: %u, slast: %u lastid: %u ver: %u, cur/nxt: %u, secnr: %u, lastnr: %u\n",
            svid, tsid, onid, segment_last_section_number, last_table_id,
            version_number, current_next, section_number,
            last_section_number);

  if (sizeStackAdvance (&s, 11, 0))
    return 1;
  ptr = sizeStackPtr (&s);

  section_length -= 11 + 4;     //header+crc
  //count number of event blocks
  p->num_events = 0;
  i = section_length;
  j = 0;
  k = 0;
  while (i > 0)
  {
    j += 10;
    k = bitsGet (ptr, j, 4, 12) + 2;    //descriptor_loop_length
    j += k;
    i -= 10;
    i -= k;
    p->num_events++;
  }
  if (!p->num_events)
  {
    debugPri (1, "num_events=0\n");
    p->events = NULL;
    return 1;
  }
  p->events = utlCalloc (p->num_events, sizeof (EVT));
  if (!p->events)
    return 1;
  i = 0;
  while (p->num_events > i)
  {
    if (get_event (&s, &(p->events[i])))
    {
      while (i--)
      {
        evtClear (&(p->events[i]));
      }
      utlFAN (p->events);
      return 1;
    }
    i++;
  }
  sizeStackPop (&s);
  if (sizeStackAdvance (&s, 0, 4))
  {
    while (i--)
    {
      evtClear (&(p->events[i]));
    }
    utlFAN (p->events);
    debugPri (1, "bound error\n");
    return 1;
  }
  ptr = sizeStackPtr (&s);
  crc = bitsGet (ptr, 0, 0, 32);
  debugPri (1, "CRC: 0x%X\n", crc);
  sizeStackPop (&s);
  p->svc_id = svid;
  p->onid = onid;
  p->tsid = tsid;
  p->segment_last_section_number = segment_last_section_number;
  p->last_table_id = last_table_id;
  p->type = eis[0];
  p->c.version = version_number;
  p->c.current = current_next;
  p->c.section = section_number;
  p->c.last_section = last_section_number;
  return 0;
}


void
clear_eis (EIS * n)
{
  int i;
  for (i = 0; i < n->num_events; i++)
  {
    evtClear (&(n->events[i]));
  }
  if (n->num_events)
    utlFAN (n->events);
  memset (n, 0, sizeof (EIS));
}

typedef struct
{
  uint8_t *sec;
  SizeStack s;
} IterEisSec;


static int
get_first_ev (IterEisSec * iter)
{
//      uint8_t *ptr;
//      uint16_t mjd;
//      uint8_t h1,m1,s1,h2,m2,s2;
//      time_t start,duration;
  uint16_t section_length;
  if (sizeStackInit (&iter->s, iter->sec) || sizeStackPush (&iter->s, MAX_SEC_SIZE))    //maximum section size
    return 1;

  section_length = bitsGet (iter->sec, 1, 4, 12);

  if (secCheckCrc (iter->sec))
  {
    return 1;
  }

  if (sizeStackAdvance (&iter->s, 3, 1))
    return 1;

  if (sizeStackPush (&iter->s, section_length))
    return 1;

  if (sizeStackAdvance (&iter->s, 11, 16))      //or 14??
    return 1;

//convert time
/*
	ptr=sizeStackPtr(&iter->s);
	mjd=bitsGet(ptr,2,0,16);
	h1=bitsDecodeBcd8(bitsGet(ptr,2,16,8));
	m1=bitsDecodeBcd8(bitsGet(ptr,2,24,8));
	s1=bitsDecodeBcd8(bitsGet(ptr,2,32,8));
	start=get_time_mjd(mjd)+get_time_hms(h1,m1,s1);
	
	h2=bitsDecodeBcd8(bitsGet(ptr,7,0,8));
	m2=bitsDecodeBcd8(bitsGet(ptr,7,8,8));
	s2=bitsDecodeBcd8(bitsGet(ptr,7,16,8));
	duration=get_time_hms(h2,m2,s2);

	*((time_t*)(ptr+2))=start;
	*((time_t*)(ptr+2+sizeof(time_t)))=duration;
	*/
  return 0;
}

static void *
iterEisSecGetCurrent (void *it)
{
  IterEisSec *iter = it;
  return sizeStackPtr (&iter->s);
}

static void
iterEisSecFirst (void *it)
{
  get_first_ev (it);
}

static int
iterEisSecAdvance (void *it)
{
  IterEisSec *iter = it;
//      uint8_t *ptr;
//      uint16_t mjd;
//      uint8_t h1,m1,s1,h2,m2,s2;
  int k;
//      time_t start,duration;
  debugPri (1, "iterEisSecAdvance\n");
  if (sizeStackAdvance (&iter->s, 10, 2))
    return 1;
  k = bitsGet (sizeStackPtr (&iter->s), 0, 4, 12) + 2;  //descriptors_loop_length
  if (sizeStackAdvance (&iter->s, k, 16))       //or 14??
    return 1;
//convert time
/*this is bullshit. tampering with the data has no effect after the section is written. Avoid it. Use time data from List elements instead.*/
/*
	ptr=sizeStackPtr(&iter->s);
	mjd=bitsGet(ptr,2,0,16);
	h1=bitsDecodeBcd8(bitsGet(ptr,2,16,8));
	m1=bitsDecodeBcd8(bitsGet(ptr,2,24,8));
	s1=bitsDecodeBcd8(bitsGet(ptr,2,32,8));
	start=get_time_mjd(mjd)+get_time_hms(h1,m1,s1);
	
	h2=bitsDecodeBcd8(bitsGet(ptr,7,0,8));
	m2=bitsDecodeBcd8(bitsGet(ptr,7,8,8));
	s2=bitsDecodeBcd8(bitsGet(ptr,7,16,8));
	duration=get_time_hms(h2,m2,s2);

	*((time_t*)(ptr+2))=start;
	*((time_t*)(ptr+2+sizeof(time_t)))=duration;
	*/
  return 0;
}

/*
static pthread_mutex_t IterEisMx=PTHREAD_MUTEX_INITIALIZER;
static ItemList IterEisAllocator;
static int IterEisAllocatorInit=0;
*/
static void
iterEisSecDestroy (void *iter)
{
  utlFAN (iter);
  /*
     pthread_mutex_lock(&IterEisMx);
     itemRelease(&IterEisAllocator,iter);
     pthread_mutex_unlock(&IterEisMx); */
}

int
iterEisSecInit (Iterator * i, uint8_t * sec)
{
  IterEisSec *d;
  memset (i, 0, sizeof (Iterator));
  /*
     pthread_mutex_lock(&IterEisMx);

     if(!IterEisAllocatorInit)
     {
     if(itemListInit(&IterEisAllocator,sizeof(IterEisSec),32))
     {
     pthread_mutex_unlock(&IterEisMx);
     return 1;
     }
     IterEisAllocatorInit=1;
     }
   */
  d = utlCalloc (1, sizeof (IterEisSec));

//      d=itemAcquire(&IterEisAllocator);
//      pthread_mutex_unlock(&IterEisMx);
  if (!d)
    return 1;
  d->sec = sec;
  if (get_first_ev (d))
  {
    utlFAN (d);
//              pthread_mutex_lock(&IterEisMx);
//              itemRelease(&IterEisAllocator,d);
//              pthread_mutex_unlock(&IterEisMx);
    return 1;
  }

  i->s.first_element = iterEisSecFirst;
  i->s.advance = iterEisSecAdvance;
  i->data = d;
  i->kind = I_SEQ_FWD;
  i->get_element = iterEisSecGetCurrent;
  i->i_destroy = iterEisSecDestroy;
  return 0;
}
